Optimize contentnode endpoint for faster browsing

bonidukic opened the pull request:
### Summary
Most of the work in this PR is focused on improving the `contentnode` endpoint by splitting it into two endpoints, creating a new, lighter option, `contentnode_slim`, to be used for content browsing.

Additional optimizations have also been added with a goal to increase the content browsing speed, such as:
- ORM prefetching to avoid additional db queries
- caching results to avoid additional db queries
- reducing the number of queries being done to get the number of nodes flagged as `coach_content`
- moved `ContentNode` progress annotation to the frontend (lazy-loaded)

### Graphs

![slim_total_requests](https://user-images.githubusercontent.com/2150991/42650490-cf6d065e-860c-11e8-825e-a301d6567dbe.png)

![slim_requests_per_s](https://user-images.githubusercontent.com/2150991/42650492-d29f7488-860c-11e8-8466-07685f1d1902.png)


### Reviewer guidance
- A couple of tests have been added to test the `contentnode_slim` endpoint, but I'm open to suggestions on how to increase test coverage if necessary
- Check if _Learn_ and _Recommended_ pages are working properly
- Keep in mind that this PR should not remove any of the existing functionalities, but strictly optimize

----

### Contributor Checklist

- [x] Contributor has fully tested the PR manually
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

bonidukic committed:
Cache result of user.roles.exists() on serializer

get_num_coach_contents method is called for each existing ContentNode
object and consequently user.roles.exists() query is executed as well -
this change simply caches the first query results on the serializer
instance to avoid duplicating queries since user instance can't change
within a single API call.

bonidukic committed:
Add contentnode_slim endpoint as lighter option

Introduce contentnode_slim endpoint and corresponding contentNodeSlim
resource which returns only the basic set of fields which are necessary
to browse the content on the frontend.

Additionally, assessmentdata key could potentially be removed but further
frontend changes are required to do so.

This change replaces calls to contentnode endpoints on the "Channels"
tab only and not on the "Recommended" tab as well.

bonidukic committed:
Separate cache key for ContentNodeSlim ancestors

Not sure if cache entries would collide with the ones from the
ContentNodeViewset, so just to make sure, let's separate these into two
separate keys, once for each of the Viewsets.

bonidukic committed:
Remove assessmentmetadata from the slim endpoint

Introduce safety check in mappers.js with an early return if the
assessmentmetada key is not present in the response, so that we can
remove that field form the endpoint, reducing the load further.

Also, add some comments to the serializers explaining their purpose.

bonidukic committed:
Do not retrieve related language data

We removed lang field from the contentnode_slim endpoint, so we can
safely remove the select_related call.

bonidukic committed:
Avoid doing extra query if file is not a thumbnail

Previously, get_storage_url would get called for each of the file
instances, now we simply check first if the file in question is even a
thumbnail, and skip that query if it's not.

NOTE: It was experimented with filtering files instances directly in the
ContentNodeSlimSerializer, but for some unknown reason response in this
context was longer than before (even though the content length was
reduced). This solution gives the best results tested so far.

bonidukic committed:
Filter out anon users before trying to cache roles

bonidukic committed:
Add semicolon to the return statement

bonidukic committed:
Slimmer contentnode endpoint on Recommended

We introduce the ContentNodeSlimResource to the Recommended page by
replacing a subset of the contentnode endpoint API calls its slimmer
counterpart.

Specifically, we do this for the following endpoint filters:
- next_steps
- popular
- resume
- recommendations_for

IMPORTANT: this change removes part of the previous functionality:
displaying user progress on the Recommended page - this will be solved
as part of this PR by adding a separate API call to fetch the user
progress.

bonidukic committed:
Prefetch root language data on channel metadata

This change is not directly related to the contentnode endpoint but
we're adding it here because it is closely connected to it.

Simply prefetch root__lang data to avoid fetching data from the DB for
each node but do it once for all of them.

Siege test report (30 concurrent users, 30s)

Before:
=======
Transactions:                    407 hits
Availability:                 100.00 %
Elapsed time:                  29.79 secs
Data transferred:               5.04 MB
Response time:                  2.10 secs
Transaction rate:              13.66 trans/sec
Throughput:                     0.17 MB/sec
Concurrency:                   28.70
Successful transactions:         407
Failed transactions:               0
Longest transaction:            5.19
Shortest transaction:           1.28

After:
======
Transactions:                    582 hits
Availability:                 100.00 %
Elapsed time:                  29.28 secs
Data transferred:               7.21 MB
Response time:                  1.47 secs
Transaction rate:              19.88 trans/sec
Throughput:                     0.25 MB/sec
Concurrency:                   29.29
Successful transactions:         582
Failed transactions:               0
Longest transaction:            4.45
Shortest transaction:           0.51

bonidukic committed:
Finally use select_related for channel metadata

bonidukic committed:
Prefetch local_file to avoid additional db lookups

Prefetching related files__local_file was previously set on the
ContentNodeViewset, but it was removed when moving to
ContentNodeSlimViewset so this change reverts that behaviour because it
was rightfully there to avoid doing additional db queries when fetching
thumbnail files urls.

bonidukic committed:
Enable coach content labels on ContentNodeSlim

In this commit we're adding back the support for labelling the coach
content for the contentnode_slim endpoints.

TODO: The get_num_coach_contents method needs to be optimized if
possible because it's producing n additional db queries (n = number
ContentNode objects) for each endpoint call.

Also, this change fixes the bug where we compared the results of the
gettatr call with the "if not" condition which of course can't work well
if the exists() method returns False.

indirectlylit commented on kolibri/core/content/management/commands/generate_schema.py:
```diff
@@ -31,6 +32,10 @@ def add_arguments(self, parser):
         parser.add_argument('version', type=str)
 
     def handle(self, *args, **options):
+        # Restrict to Python 3 due to issues with `json.dump` on Python 2
```
what sort of issues?

bonidukic replied:
```
(kolibri)  boni@t470s  ~/projects/learning-equality/kolibri  (feature/optimize-contentnode-endpoint) >  kolibri manage generate_schema 3
INFO     Option RUN_MODE in section [Deployment] being overridden by environment variable KOLIBRI_RUN_MODE
/home/boni/.venvs/kolibri/local/lib/python2.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
Installed 25 object(s) from 1 fixture(s)
Traceback (most recent call last):
  File "/home/boni/.venvs/kolibri/bin/kolibri", line 9, in <module>
    load_entry_point('kolibri', 'console_scripts', 'kolibri')()
  File "/home/boni/projects/learning-equality/kolibri/kolibri/utils/cli.py", line 631, in main
    manage(command, args=django_args)
  File "/home/boni/projects/learning-equality/kolibri/kolibri/utils/cli.py", line 421, in manage
    execute_from_command_line(argv=argv)
  File "/home/boni/projects/learning-equality/kolibri/kolibri/dist/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/home/boni/projects/learning-equality/kolibri/kolibri/dist/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/boni/projects/learning-equality/kolibri/kolibri/dist/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/boni/projects/learning-equality/kolibri/kolibri/dist/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/home/boni/projects/learning-equality/kolibri/kolibri/core/content/management/commands/generate_schema.py", line 65, in handle
    json.dump(data, f)
  File "/usr/lib/python2.7/json/__init__.py", line 190, in dump
    fp.write(chunk)
TypeError: write() argument 1 must be unicode, not str
```

bonidukic replied:
@indirectlylit FYI I discussed this with @rtibbles and we agreed that since this is internal command that it might be easiest to require Python 3 - but if you feel we should support Python 2 as well here I can play with this to figure out something that works?

bonidukic replied:
I found this SO answer helpful earlier: https://stackoverflow.com/a/36008538/6817428

indirectlylit replied:
Main issue is that currently builds _only_ work in py2, so that leaves us with an inconsistent dev experience: https://github.com/learningequality/kolibri/blob/develop/Makefile#L122

@benjaoming do you recall why py2 is necessary for building?

benjaoming replied:
Yes, we only build on Python 2. That's the only way currently to have a Py2+3 build.

So we have to have everything py2+3, there is no nice way to handle exceptions to the rule IMO.

bonidukic replied:
I mean, it's not like I changed any previous behaviour here, this wasn't working on Python 2 anyway, I just added a message to the user to avoid them having to figure out what's going on.

benjaoming replied:
Oh right okay, then let's just say that someone at some point in time will have to find out that this `generate_schema` doesn't work on python 2 and fix that if it's a problem :)

bonidukic replied:
Yep, I'll try to take a look at it within this PR, AFAIK there is a difference between using `json.dumps()` and `json.dump()` with respect to the `ensure_ascii` (among other things) so there might be an easy solution.

bonidukic replied:
FYI: I removed this commit from this PR and created a separate PR with a compatibility fix: https://github.com/learningequality/kolibri/pull/4020

bonidukic committed:
Avoid descendant queries for node which are not CC

Previously we've been doing recursive descendants queries for all nodes
regardless of the fact that they're not coach content.

This change adds an early 0 value return if the passed content is
not a coach content, because due to the way in which the content
annotation works, its descendants have to be non-coach content as well.

bonidukic committed:
Annotate progress on nodes also (not only topics)

Since we removed contentnode progress annotation from the
contentnodeslim endpoint, we expanded the existing functionality of
anontating progress on topics to include nodes as well, as that was
missing.

NOTE: This is related to the Learn module, progress annotation is to be
added to the Recommended module as well.

rtibbles commented on kolibri/core/content/api.py:
```diff
@@ -473,6 +475,62 @@ def all_content(self, request, **kwargs):
+
+        cache.set(cache_key, ancestors, 60 * 10)
+
+        return Response(ancestors)
```
There are probably other list and detail endpoints from the ContentNode viewset that can be ported to ContentNodeSlim - but let's get this fixed and merged and then iterate!

rtibbles commented:
Test errors are one Python linting error, and one testing error that probably only appeared on Postgresql due to ordering issues (as we were only making an assertion about the first item in the response).

bonidukic committed:
Annotate contentnode progress on Recommended page

Since we're using contentnodeslim endpoint on the Recommended page which
no longer returns progress data, we're annotating the nodes manually
by lazily loading that information from the frontend.

bonidukic committed:
Add tests for contentnode_slim endpoint

rtibbles commented on kolibri/core/content/api.py:
```diff
@@ -473,6 +475,62 @@ def all_content(self, request, **kwargs):
+        obj = get_object_or_404(queryset, **filter_kwargs)
+
+        # May raise a permission denied
+        self.check_object_permissions(self.request, obj)
```
We can not bother with permission checks on ContentNode, as they have no permissions on them.

bonidukic replied:
I'm guessing that in that case we want to remove the `check_object_permissions` check on both `ContentNodeViewset` and `ContentNodeSlimViewset`?

rtibbles replied:
Yeah, we could probably have a 'non-permissioned' Viewset base class.

rtibbles replied:
Might be other things we could exclude too.

rtibbles replied:
Can manage this in future cleanup.

rtibbles commented on kolibri/plugins/learn/assets/src/state/mutations/index.js:
```diff
@@ -35,12 +35,22 @@ export default {
     });
   },
+  SET_RECOMMENDED_NODES_PROGRESS(state, progressArray) {
+    Object.keys(state.pageState).forEach(function(key) {
```
Might want to make this explicit and use `['recommended', 'popular', 'nextSteps']` instead of reading the object keys of the `pageState`.

bonidukic replied:
I was thinking about that but wasn't sure about the general verbosity philosophy so I went with this, but will gladly update as per your suggestion.

rtibbles replied:
My main reason for explicitness is just in case there are other keys on the pageState that aren't one of these three.

bonidukic replied:
Yes, I understand completely, my thinking was that if another key was to be added to `pageState` it would be another set of _Recommended_ nodes which we'd want to annotate anyway - but I agree that listing them explicitly is definitely safer and someone will have to add it to the mutation manually if they see that it's not working rather than causing regression if the key added to `pageState` wasn't a list of nodes.

rtibbles reviewed (changes requested):
One small code tweak. Will be manually testing also.

rtibbles commented:
Manual testing looks good to me.

rtibbles committed:
Fix tests, linting, and cleanup mutation.

rtibbles reviewed (approved):
Updates made, manual testing works!
